src/parser/Parser.ts

```diff
@@ -2718,6 +2730,138 @@ export class Parser {
+            currentToken = previousTokenResult?.token;
+            currentTokenIndex = previousTokenResult?.index;
+        }
+        isUnknown = isUnknown || previousTokenResult?.isUnknown;
+        if (isUnknown || this.isAcceptableChainToken(currentToken)) {
+            return { token: currentToken, index: currentTokenIndex, isUnknown: isUnknown };
+        }
+        return undefined;
+    }
+    private isAcceptableChainToken(currentToken: Token, lastTokenHasWhitespace = false): boolean {
+        if (!currentToken || lastTokenHasWhitespace) {
+            return false;
+        }
+        if (currentToken.kind === TokenKind.Identifier) {
+            return true;
+        }
+        if (currentToken.leadingWhitespace.length === 0) {
+            // start of the chain
+            return AllowedLocalIdentifiers.includes(currentToken.kind);
+        }
+        // not the start of the chain
+        return AllowedProperties.includes(currentToken.kind);
+    }
+    /**
+     * Builds up a chain of tokens, starting with the first in the chain, and ending with currentToken
+     * e.g. m.prop.method().field (with 'field' as currentToken) -> ["m", "prop", "method", "field"], with each element as a token
+     * @param currentToken the token that is the end of the chain
+     * @returns array of tokens
+     */
+    public getTokenChain(currentToken: Token): TokenChain {
+        const tokenChain: Token[] = [];
+        let currentTokenIndex = this.tokens.indexOf(currentToken);
+        let previousTokenResult: TokenWithIndex;
+        let lastTokenHasWhitespace = false;
+        let includesUnknown = false;
+        previousTokenResult = this.getPreviousTokenInChain(currentTokenIndex, true);
+        currentToken = previousTokenResult?.token;
+        currentTokenIndex = previousTokenResult?.index;
+        if (this.isAcceptableChainToken(currentToken)) {
+            tokenChain.push(currentToken);
+            lastTokenHasWhitespace = currentToken?.leadingWhitespace.length > 0;
+        }
+        if (!lastTokenHasWhitespace) {
+            previousTokenResult = this.getPreviousTokenInChain(currentTokenIndex);
+            currentToken = previousTokenResult?.token;
+            currentTokenIndex = previousTokenResult?.index;
+            includesUnknown = !!previousTokenResult?.isUnknown;
+            while (!includesUnknown && this.isAcceptableChainToken(currentToken, lastTokenHasWhitespace)) {
+                tokenChain.push(currentToken);
+                lastTokenHasWhitespace = currentToken?.leadingWhitespace.length > 0;
+                if (!lastTokenHasWhitespace) {
+                    previousTokenResult = this.getPreviousTokenInChain(currentTokenIndex);
+                    currentToken = previousTokenResult?.token;
+                    currentTokenIndex = previousTokenResult?.index;
+                    includesUnknown = previousTokenResult?.isUnknown;
```

Comment:
It seems like this gets reset on each token. Shouldn't you be doing something like this to flag if the chain has at least one unknown?
```typescript
includesUnknown = includesUnkown || previousTokenResult?.isUnknown;
```

Reply:
As soon as it includes an unknown token, then the loop exits....